fe/fe-core/src/main/java/com/starrocks/external/hive/events/MetastoreEventsProcessor.java

```diff
@@ -0,0 +1,409 @@
+            tablesLock.writeLock().unlock();
+        }
+    }
+
+    public HiveTable getHiveTable(String resourceName, String dbName, String tblName) {
+        tablesLock.readLock().lock();
+        try {
+            return tables.get(resourceName, getTablesColumnKey(dbName, tblName));
+        } finally {
+            tablesLock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Gets metastore notification events from the given eventId. The returned list of
+     * NotificationEvents are filtered using the NotificationFilter provided if it is not null.
+     * @param resourceName The resource name of current hive metastore instance.
+     * @param getAllEvents If this is true all the events since eventId are returned.
+     *                     Note that Hive MetaStore can limit the response to a specific
+     *                     maximum number of limit based on the value of configuration
+     *                     {@code hive.metastore.max.event.response}.
+     *                     If it is false, only {@link Config#hms_events_batch_size_per_rpc} events are
+     *                     returned, caller is expected to issue more calls to this method
+     *                     to fetch the remaining events.
+     * @param filter This is a nullable argument. If not null, the events are filtered
+     *               and then returned using this. Otherwise, all the events are returned.
+     * @return List of NotificationEvents from metastore since eventId.
+     * @throws MetastoreNotificationFetchException In case of exceptions from HMS.
+     */
+    private List<NotificationEvent> getNextHMSEvents(String resourceName,
+                                                    final boolean getAllEvents,
+                                                    @Nullable final IMetaStoreClient.NotificationFilter filter)
+            throws MetastoreNotificationFetchException {
+        Long lastSyncedEventId = null;
+        try {
+            HiveMetaClient client = hiveRepository.getClient(resourceName);
+            if (client == null) {
+                return Collections.emptyList();
+            }
+
+            lastSyncedEventId = lastSyncedEventIds.get(resourceName);
+            // restart fe or just created hive table.
+            if (lastSyncedEventId == null) {
+                lastSyncedEventIds.put(resourceName, client.getBaseHmsEventId());
+                return Collections.emptyList();
+            }
+
+            CurrentNotificationEventId currentNotificationEventId = client.getCurrentNotificationEventId();
+            long currentEventId = currentNotificationEventId.getEventId();
+            if (currentEventId == lastSyncedEventId) {
+                return Collections.emptyList();
+            }
+
+            int batchSize = getAllEvents ? -1 : Config.hms_events_batch_size_per_rpc;
+            NotificationEventResponse response = client.getNextNotification(lastSyncedEventId, batchSize, null);
```

Comment:
Can useless events be filtered when client is getting next notification, so that less data need to be send.

Reply:
There are currently no parameters to filter on the hms side.

Reply:
It is true that a filter should be passed in, but the old version of hive cannot pass in the filter. In order to be compatible with multiple versions, we currently do not support the use of EventFilter.